### ISSUE TITLE
BSD Firewall module error?

### ISSUE THREAD

**Issue by crlmahlberg** [webmin/webmin#1327]
I am running FreeBSD 12.1 on a DigitalOcean VPS (a "droplet") and want to use _ipfw_. The firewall itself seems to work, but Webmin's _BSD Firewall_ module gives me the following error (Webmin version is 1.962 and Perl version 5.32):

> An error occurred fetching the active firewall rules using the command /sbin/ipfw list :
> 
> . Maybe the kernel has not been configured to support firewalling, or the module configuration is incorrect.
> 

`sudo /sbin/ipfw` gives me the following output, so the command actually seems to work:

> 00100 allow ip from any to any via lo0
> 00200 deny ip from any to 127.0.0.0/8
> 00300 deny ip from 127.0.0.0/8 to any
> 00400 deny ip from any to ::1
> 00500 deny ip from ::1 to any
> 00600 allow ipv6-icmp from :: to ff02::/16
> 00700 allow ipv6-icmp from fe80::/10 to fe80::/10
> 00800 allow ipv6-icmp from fe80::/10 to ff02::/16
> 00900 allow ipv6-icmp from any to any icmp6types 1
> 01000 allow ipv6-icmp from any to any icmp6types 2,135,136
> 65000 allow ip from any to any
> 65535 deny ip from any to any

Here is my complete _rc.conf_ file used to setup _ipfw_:

```
zfs_enable="YES"
sshd_enable="YES"
hostname="****.****.*******.**"
defaultrouter="***.***.***.*"
ifconfig_vtnet0="inet ***.***.***.** netmask 255.255.240.0"
ifconfig_vtnet1="inet 10.135.0.3 netmask 255.255.0.0"
ifconfig_vtnet0_alias0="inet 10.19.0.5 netmask 255.255.0.0"
ifconfig_vtnet0_ipv6="inet6 2A03:B0C0:0003:00E0:0000:0000:000D:8001 prefixlen 64"
ipv6_activate_all_interfaces="yes"
ipv6_defaultrouter="2a03:b0c0:0003:00e0:0000:0000:0000:0001"
digitalocean="YES"
digitaloceanpre="YES"
cloudinit_enable="NO"
#gateway_enable="YES"
#natd_enable="YES"
#natd_interface="vtnet0"
firewall_enable="YES"
firewall_type="open"
firewall_quiet="NO"
firewall_logging="YES"
#firewall_myservices="22/tcp 80/tcp 443/tcp 10000/tcp"
#firewall_allowservices="any"
#firewall_script="/etc/rc.firewall"
#firewall_script="/etc/ipfw.rules"
#firewall_type="/etc/ipfw.rules"
#firewall_myservices="22 80 443 10000"
#firewall_logdeny="YES"
#firewall_nat_enable="YES"
proftpd_enable="YES"
webmin_enable="YES"
monit_enable="YES"
acme_dns_enable="YES"
syncthing_enable="YES"
clamav_freshclam_enable="YES"
nginx_enable="YES"
php_fpm_enable="YES"
postgresql_enable="YES"
pgadmin4_enable="YES"
redis_enable="YES"
znc_enable="YES"
znc_user="znc"
znc_conf_dir="/usr/local/etc/znc"

# DigitalOcean Dynamic Configuration lines and the immediate line below it,
# are removed each boot.

# DigitalOcean Dynamic Configuration
defaultrouter="***.***.***.*"
# DigitalOcean Dynamic Configuration
ifconfig_vtnet0="inet ***.***.***.** netmask 255.255.240.0"
# DigitalOcean Dynamic Configuration
ifconfig_vtnet0_alias0="inet 10.19.0.5 netmask 255.255.0.0"
# DigitalOcean Dynamic Configuration
ifconfig_vtnet1="inet 10.135.0.3 netmask 255.255.0.0"
# DigitalOcean Dynamic Configuration
ifconfig_vtnet0_ipv6="inet6 2A03:B0C0:0003:00E0:0000:0000:000D:8001 prefixlen 64"
# DigitalOcean Dynamic Configuration
ipv6_defaultrouter="2a03:b0c0:0003:00e0:0000:0000:0000:0001"
```

I tried making it work with a manual _ipfw.rules_ file, but that didn't make any difference. I did not modify _/etc/rc.firewall_ in any way.
What is the issue here? So far I have never had any problems with Webmin, all the modules work without a problem, except this one. Thanks for helping me resolve this issue!

**Comment by bsiege** [comment]
I think you have a mix between pf and ipfw which both have a module. I recommend ipfw.
[FreeBSD Firewalls](https://www.freebsd.org/doc/en/books/handbook/firewalls.html)

**Comment by jcameron** [comment]
What output do you get if you run `/sbin/ipfw list` as `root` at the command line?

**Comment by bsiege** [comment]
Hi,
can confirm that there something changed in this module. Have a quite complex ruleset. Last time touched in August19 with webmin. Now i see same problem like @crlmahlberg . Will debug tonight.

**Comment by crlmahlberg** [comment]
@jcameron Here is the output:

> 00010 allow ip from any to any via lo0
00020 deny ip from any to 127.0.0.0/8
00030 deny ip from 127.0.0.0/8 to any
00040 deny tcp from any to any frag
00050 check-state :default
00060 allow tcp from any to any established
00070 allow ip from any to any out keep-state :default
00080 allow icmp from any to any
00090 allow ipv6-icmp from any to any
00100 allow tcp from any to any 21 in setup keep-state :default
00110 allow tcp from any to any 60000-65535 in setup keep-state :default
00120 allow tcp from any to any 22 in setup keep-state :default
00130 allow tcp from any to any 53 in keep-state :default
00140 allow udp from any to any 53 in keep-state :default
00150 allow tcp from any to any 80 in setup keep-state :default
00160 allow tcp from any to any 443 in setup keep-state :default
00170 deny log ip from any to any
65535 deny ip from any to any

My firewall is now set up to use a very basic _/etc/ipfw.rules_ file.

**Comment by bsiege** [comment]
```
/usr/local/lib/webmin-1.962/ipfw] # tail -f /var/log/webmin/webmin.debug 
80082 [29/Dec/2020 17:01:35.848973] root my_clients_ipv6_address - READ "/usr/local/lib/webmin-1.962/authentic-theme/config"
80082 [29/Dec/2020 17:01:35.860713] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/custom-lang"
80082 [29/Dec/2020 17:01:35.865774] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/ipfw/custom-lang"
80082 [29/Dec/2020 17:01:35.866891] root my_clients_ipv6_address - READ "/var/log/webmin/module.infos.cache"
80082 [29/Dec/2020 17:01:35.882284] root my_clients_ipv6_address - READ "/usr/local/lib/webmin-1.962//defaultacl"
80082 [29/Dec/2020 17:01:35.883013] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/root.acl"
80082 [29/Dec/2020 17:01:35.883818] root my_clients_ipv6_address - READ "/usr/local/etc/webmin//root.acl"
80082 [29/Dec/2020 17:01:35.884537] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/installed.cache"
80082 [29/Dec/2020 17:01:35.886532] root my_clients_ipv6_address - READ "/usr/local/lib/webmin-1.962/ipfw/module.info"
80082 [29/Dec/2020 17:01:35.887350] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/webmin.cats"
80082 [29/Dec/2020 17:01:35.887849] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/webmin.descs"
80082 [29/Dec/2020 17:01:35.891359] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/net/module.info"
80082 [29/Dec/2020 17:01:35.903253] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/net/config"
80082 [29/Dec/2020 17:01:35.905455] root my_clients_ipv6_address - READ "/usr/local/etc/webmin/net/custom-lang"
80082 [29/Dec/2020 17:01:35.915403] root my_clients_ipv6_address net READ "/usr/local/lib/webmin-1.962/net/defaultacl"
80082 [29/Dec/2020 17:01:35.916205] root my_clients_ipv6_address net READ "/usr/local/etc/webmin/net/root.acl"
80082 [29/Dec/2020 17:01:35.932281] root my_clients_ipv6_address net READ "/etc/defaults/rc.conf"
80082 [29/Dec/2020 17:01:35.940512] root my_clients_ipv6_address net READ "/etc/rc.conf"
80082 [29/Dec/2020 17:01:36.094405] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/authentic-theme/config"
80082 [29/Dec/2020 17:01:36.095166] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/authentic-theme/settings.js"
80082 [29/Dec/2020 17:01:36.096407] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/authentic-theme/settings-root"
80082 [29/Dec/2020 17:01:36.099153] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/authentic-theme/custom-lang"
80082 [29/Dec/2020 17:01:36.103823] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/server-manager/module.info"
80082 [29/Dec/2020 17:01:36.104310] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/virtual-server/module.info"
80082 [29/Dec/2020 17:01:36.104850] root my_clients_ipv6_address ipfw READ "/usr/local/etc/usermin/version"
80082 [29/Dec/2020 17:01:36.105379] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/ipfw/defaultacl"
80082 [29/Dec/2020 17:01:36.105919] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/ipfw/root.acl"
80082 [29/Dec/2020 17:01:36.106367] root my_clients_ipv6_address ipfw CMD "cmd=hostname"
80082 [29/Dec/2020 17:01:36.116317] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/shell/module.info"
80082 [29/Dec/2020 17:01:36.118141] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/webmin/module.info"
80082 [29/Dec/2020 17:01:36.119922] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/system-status/module.info"
80082 [29/Dec/2020 17:01:36.121558] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/package-updates/module.info"
80082 [29/Dec/2020 17:01:36.122867] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/csf/module.info"
80082 [29/Dec/2020 17:01:36.123460] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/authentic-theme/theme.info"
80082 [29/Dec/2020 17:01:36.124215] root my_clients_ipv6_address ipfw READ "/usr/local/lib/webmin-1.962/authentic-theme/theme.info"
80082 [29/Dec/2020 17:01:36.127441] root my_clients_ipv6_address ipfw READ "/tmp/.webmin/.theme_NzQ4ZjMxNzBhYjJk_webmin_redirected_root"
80082 [29/Dec/2020 17:01:36.128087] root my_clients_ipv6_address ipfw READ "/usr/local/etc/webmin/virtual-server/config"
80082 [29/Dec/2020 17:01:36.128900] root my_clients_ipv6_address ipfw CMD "cmd=date +"%z""
80082 [29/Dec/2020 17:01:36.145749] root my_clients_ipv6_address - STOP "runtime=1"
```

Found no obvious problem in webmin.debug except the fact that "/usr/local/etc/webmin/virtual-server/config" does not exist?

**Comment by jcameron** [comment]
@crlmahlberg apologies, I gave the wrong command.

What does `ipfw show` output?

**Comment by crlmahlberg** [comment]
@jcameron `/sbin/ipfw show` run as `root` give me

![Bildschirmfoto vom 2020-12-30 07-45-18](https://user-images.githubusercontent.com/54212019/103334799-0aee0e80-4a73-11eb-8f26-dbbae6d5af74.png)

**Comment by jcameron** [comment]
Ok I found a bug that can cause this error about `ipfw list` to be incorrectly displayed. It will be fixed in the upcoming 1.970 Webmin release.

**Comment by crlmahlberg** [comment]
@jcameron Great, looking forward to the upgrade. Thanks.
